drivers/ili9341/ili9341_disp_dev.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (C) 2019 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_ili9341
+ * @{
+ *
+ * @file
+ * @brief       Driver adaption to disp_dev generic interface
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ * @}
+ */
+
+#include <stdint.h>
+#include <errno.h>
+
+#include "ili9341.h"
+#include "ili9341_disp_dev.h"
+
+static void _ili9341_map(disp_dev_t *dev, uint16_t x1, uint16_t x2,
+                  uint16_t y1, uint16_t y2, const uint16_t *color)
+{
+    ili9341_t *ili9341 = (ili9341_t *)dev;
+    ili9341_pixmap(ili9341, x1, x2, y1, y2, color);
+}
+
+static int _ili9341_get(disp_dev_t *disp_dev, disp_opt_t opt, void *value, size_t max_len)
```

Comment:
I wish we had an easy way to compare this mechanism with a corresponding function pointer based version.

I mean, if there'd be inline wrappers like:

```
static int disp_dev_get_max_width(dev)
{
  if !(dev->ops->get_max_width) {
    return -ENOTSUP;
  }
  return dev->ops->get_max_width(dev);
}
```
(this would be shared by all).

As the ops struct is constant, only supported functions would need to be mentioned in this:

```
const disp_dev_driver_t ili9341_disp_dev_driver = {
    .map = _ili9341_map,
    .get = _ili9341_get,
    .set = _ili9341_set,
};
```

I'd assume that the code would be of similar size, but always fully type checked.
(Unless we espect *a lot* of not implemented operations.)

Reply:
Also, if a function is implemented (by implementing the corresponding `static int _ili9341_scroll_down()` or whatever, the compiler comlplains about an unused function if it is not added to the ops struct.

Reply:
So if I understand correctly you want to test it against something like:
```
const disp_dev_driver_t ili9341_disp_dev_driver = {
    .map = _ili9341_map,
    .get_width = _ili9341_get_width,
    .get_heigth = _ili9341_get_heigth,
    .set_brightness = _ili9341_set_brightness,
};
```
?

Reply:
Yes, thats what I mean. I know we use this in netdev, but there it is also meant to work over netapi / msgs.

I'll try to sketch it, to get an idea of the overhead of an implemented / unimplemented function.

Reply:
I would prefer to keep the ops struct as it is implemented in the PR and add wrapper functions around each of them.
I see a benefit in API usage. e.g.
```
disp_dev_map(dev, ...)
disp_dev_get(dev, DISP_OPT_<option name>, ...)
disp_dev_set(dev, DISP_OPT_<option name>, ...)
```
instead of:
```
dev->driver->map(dev, ...)
dev->driver->get(dev, DISP_OPT_<option name>, ...)
dev->driver->set(dev, DISP_OPT_<option name>, ...)
```

Reply:
I had this idea in mind already, but just forgot to implement it.

Reply:
> `disp_dev_get(dev, DISP_OPT_<option name>, ...)`
> `disp_dev_set(dev, DISP_OPT_<option name>, ...)`

I think the detail is in the `...`. As proposed, this would be:

`disp_dev_get(disp_dev_t *disp_dev, disp_opt_t opt, void *value, size_t max_len);`

The user needs to manually typecheck the pointer and max_len, or check the result. Sure, the asserts catch that, but it is clumsy.

E.g., getting `DISP_OPT_MAX_WIDTH`:
```
int max_with;
int res = disp_dev_get(dev, DISP_OPT_MAX_WIDTH, &max_width, sizeof(max_width));
```
Do you know if this works as expected on a 32bit platform? The compiler won't check any type.

That can surely be wrapped:

```
static inline uint16_t disp_dev_get_max_width(dev)
{
    int16_t max_width;
    int res = disp_dev_get(dev, DISP_OPT_MAX_WIDTH, &max_width, sizeof(max_width));
   if (res) return 0;
   return max_width;
}
```

... but at that point my proposal is much less hassle.
    
Again, using pointers & max_len was a kludge in netdev, in order to allow calling the API directly over netapi / messages.

Reply:
I don't know how this would affect code size. I could imagine that having an extra function for returning a single value might be a waste. Then again, with a static inline getter, that can easily be changed to be a static field, for common values. (`static inline disp_dev_get_max_width(dev) return dev->params->max_width;`).

@aabadie do you see my point? I can try changing the branch towards fully function pointer based, to get an idea of the code size implications.

Reply:
> Do you know if this works as expected on a 32bit platform? The compiler won't check any type.

I'm not sure to understand what you mean. Is it because the example you provide uses `int` instead of uint16_t ?

The problem with specific wrappers is that it contraints the available common functions. And not all display device will provide `get/set_brightness`, `get/set_color_depth`, etc. The only common minimum is about width/height attributes.

If you agree and if I understand it correctly, I can try your proposed approach, and give some comparison.

Reply:
> I'm not sure to understand what you mean. Is it because the example you provide uses `int` instead of uint16_t ?

yes.



> The problem with specific wrappers is that it contraints the available common functions. 

This I don't understand.

> And not all display device will provide `get/set_brightness`, `get/set_color_depth`, etc. The only common minimum is about width/height attributes.

The price for a not supported function is the size of one pointer, in flash. Unless we'd have a high ratio of those, it could be acceptable.

> If you agree and if I understand it correctly, I can try your proposed approach, and give some comparison.

If you factor out the functions of the current switch-based approach, it could be easy to switch.

e.g.:

```
switch (option):
    case DISP_DEP_GET_FOO:
      ...
      ((uint16_t*)ptr) = dev->foo;
    break;
```

->

```

static uint16_t _get_foo(dev) {
  return dev->foo;
}

switch (option) {
   case DISP_DEP_GET_FOO:
      ...
      ((uint16_t*)ptr) = _get_foo(dev);
    break;
    ...
```

Up to here it should actually result in the same code, but now `_get_foo()` can directly be used in the drivers struct.

Reply:
@kaspar030, I tried your proposal and it's 60B smaller than the current one. I'll push the changes in a moment.